### ISSUE TITLE
st/encode and st/conform with s/double-in and s/int-in inside data specs stopped working as of 0.9.0+

### ISSUE THREAD

**Issue by alzadude** [metosin/spec-tools#184]
With versions 0.8.2, 0.8.3 the following data spec is successfully string encoded (and decoded) as expected:
```
(def double-in-spec
  (std/spec
   {:name ::double-in
    :spec {:a-double (s/double-in :min 0 :NaN? false :infinite? false)}}))
```
```
(st/encode double-in-spec (gen/generate (s/gen double-in-spec)) st/string-transformer)
; {:a-double "114.0"}
```
```
(st/conform double-in-spec {:a-double "114.0"} st/string-transformer)
; {:a-double 114.0}
```
But with version 0.9.0+ (0.9.0, 0.9.1), it doesn't seem to work as expected any more? As in, the value is not encoded (or decoded):
```
(st/encode double-in-spec (gen/generate (s/gen double-in-spec)) st/string-transformer)
; {:a-double 225.30078125}
```
```
(st/conform double-in-spec {:a-double "225.30078125"} st/string-transformer)
; :clojure.spec.alpha/invalid
```
I've been relying on this in a project, and I just tried to upgrade (trying to keep up-to-date) and unexpectedly hit this issue... 😕

Additionally it seems `s/int-in` is also affected in the same way...

**Comment by miikka** [comment]
Hi, thanks for reporting this. I ran a bisect and looks like it was commit https://github.com/metosin/spec-tools/commit/fa816df04f481eaa5881f843cf1c20b3aed70b34 that introduced the bug. (cc. @ikitommi)

**Comment by alzadude** [comment]
> Hi, thanks for reporting this. I ran a bisect and looks like it was commit fa816df that introduced the bug. (cc. @ikitommi)

Hi there, I just found that `st/conform` fails in the same way for 0.9.0+ (it works fine with 0.8.3).

I'll update the title and description.

Note: In contrast to `st/coerce`, `st/decode` seems to work ok for the spec in the description.. However I'm not using `st/decode`.. Instead I'm using `st/conform` followed by `s/unform`, because I have `s/cat` and regex ops in my data specs, and it appears that `st/coerce` (and therefore also `st/decode`) has an issue with s/cat specs: https://github.com/metosin/spec-tools/issues/149

**Comment by ikitommi** [comment]
Checked what is happening here. The root cause seems to be that `st/conform` doesn't understand composite types, there are no tests for it I guess. You can use either `st/coerce` or `st/decode` to get those working, but will try to fix the `st/conform` too.

```clj
(parse/parse-spec (s/double-in :min 0 :NaN? false :infinite? false))
; {:type [:and [:double]],
;  :spec-tools.parse/items [{:spec #object[clojure.core$double_QMARK_ 0x396341f9 "clojure.core$double_QMARK_@396341f9"],
;                            :type :double}
;                           {:type nil}
;                           {:type nil}
;                           {:type nil}]}

(s/form (s/double-in :min 0 :NaN? false :infinite? false))
;(clojure.spec.alpha/and
;  clojure.core/double?
;  (clojure.core/fn [%] (clojure.core/not (Double/isInfinite %)))
;  (clojure.core/fn [%] (clojure.core/not (Double/isNaN %)))
;  (clojure.core/fn [%] (clojure.core/<= 0 %)))

(st/decode double-in-spec {:a-double "114.0"} st/string-transformer)
; {:a-double 114.0}

(st/coerce double-in-spec {:a-double "225.30078125"} st/string-transformer)
; {:a-double 225.30078125}
```

**Comment by ikitommi** [comment]
Fixed in #185

**Comment by ikitommi** [comment]
published in `[metosin/spec-tools "0.9.2-alpha2"]`

**Comment by alzadude** [comment]
> published in [metosin/spec-tools "0.9.2-alpha2"]

Thanks! I've updated to 0.9.2 and it's looking good 👍
